app/scripts/migrations/163.ts

```diff
@@ -0,0 +1,91 @@
+import { cloneDeep } from 'lodash';
+import { hasProperty, isObject } from '@metamask/utils';
+
+type VersionedData = {
+  meta: { version: number };
+  data: Record<string, unknown>;
+};
+
+export const version = 163;
+
+/**
+ * This migration removes the `tokens`, `detectedTokens`, and `ignoredTokens` properties from the TokensController state for users who do not have tokenListController state.
+ *
+ * If the TokensController is not found or is not an object, the migration logs an error,
+ * but otherwise leaves the state unchanged.
+ *
+ * @param originalVersionedData - The versioned extension state.
+ * @returns The updated versioned extension state without the tokens property.
+ */
+export async function migrate(
+  originalVersionedData: VersionedData,
+): Promise<VersionedData> {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+
+  versionedData.data = transformState(versionedData.data);
+
+  return versionedData;
+}
+
+function transformState(
+  state: Record<string, unknown>,
+): Record<string, unknown> {
+  // If property TokensController is not present, only log a warning and return the original state.
+  if (!hasProperty(state, 'TokensController')) {
+    console.warn(`newState.TokensController is not present`);
+    return state;
+  }
+
+  const tokensControllerState = state.TokensController;
+
+  // If property tokensControllerState is there but not an object, capture a sentry error and return state
+  if (!isObject(tokensControllerState)) {
+    global.sentry?.captureException?.(
+      new Error(
+        `Migration ${version}: TokensController is type '${typeof tokensControllerState}', expected object.`,
+      ),
+    );
+    return state;
+  }
+
+  // If property TokenListController is not present, only log a warning and return the original state.
+  if (!hasProperty(state, 'TokenListController')) {
+    console.warn(`newState.TokenListController is not present`);
+    return state;
```

Comment:
I thought the purpose of this migration was to delete the relevant `tokensControllerState` properties for installations where `TokenListController` is missing? If we're bailing here, then this migration seems to serve no purpose

Reply:
I think this is where i was confused; My understanding was that for users with old installations and no `tokenListController` state nor `tokensController` state we would want to `console.warn` instead of captureException.
Which is true; but i think my logic does not cover the case where users do not have TokenListController but still have tokensController state; in which case we want to cleanup state.

Reply:
Unless i'm misunderstanding something again ^^"; i think when you said
`Assuming I'm on the same page about the purpose of reintroducing this, then I think we could delete the TokenListController steps completely`
You meant
`Assuming I'm on the same page about the purpose of reintroducing this, then I think we could delete the TokensController steps completely`
Since we only care about the case where user does not have `tokenListController` state but still has `tokensController` state;

Reply:
Oh my bad, i missed the fact that even if tokenListController does not exist it would get initialized when users upgrade to v12.18.x.
Alright almost there :p pushed a new update here [76c6bcc](https://github.com/MetaMask/metamask-extension/pull/33439/commits/76c6bcc62731a9223cfdc9464782e92f29538b98) 🙏 ^^